grammar.go

```diff
@@ -22,6 +22,38 @@ func newGeneratorContext(lex lexer.Definition) *generatorContext {
 	}
 }
 
+func (g *generatorContext) addUnionDefs(defs []unionDef) error {
+	unionNodes := make([]*union, len(defs))
+	for i, def := range defs {
```

Comment:
Unions may rely on one another, so we have to add them all to the `typeNodes` map before we start processing their members, and then update the nodes in-place.